### PR TITLE
✨ Async Publishing (#1000)

### PR DESCRIPTION
**Pull request by mistryrn**
In lieu of a detailed PR description, I've added PR comments on each file (where necessary).

The TL;DR is we've essentially duplicated the async "GDC Variants Importer" and repurposed it as a new async "Publisher" service. This new "Publisher" uses the same queue system, has its own `/publish` router with similar endpoints to the importer, but publishes models to ES instead of importing variants for them.

On the UI we've replaced all calls to the old publish endpoints with calls to the new async endpoints, removed any locking/freezing of the UI when publishing, added a "Publish Progress" banner similar to the one used for GDC import progress, and updated the ModelStatusPill component to display "Publish in Progress" for models that are currently being published.

Due to time constraints, there is quite a bit of code duplication unfortunately. A few import-specific functions/components have been duplicated and had some parts removed and the rest renamed from "import____" to "publish____". A nice refactor/cleanup ticket can be made in the future, but this works for now.

# 🚨 RELEASE INSTRUCTIONS 🚨 #
Once this PR and the HHS Vulnerability Disclosure PR (https://github.com/nci-hcmi-catalog/portal/pull/1010) are merged to `develop`, please do the following:
1. Make and merge a "🔖  Release 1.17.0" PR from `develop` into `master`
2. Make a new Release, tag "1.17.0" off of `master`
3. Provide the following instructions to Cuong for deployment: https://docs.google.com/document/d/1kTUCJmAX5HUygZ4oGdAN3-sZDIV1YrgbD4KxEfPtwp0/edit#

---

### Commits

**♻️ Refactor Individual Model Publishing on Model Edit Page (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Change the `Publish` button on the Model Edit page to use the `/action/publish/:modelName` endpoint instead of re-using the same endpoint as the `Save` button but with an extra flag set

**✨ Add New Publish Router (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Adds a new `/publish` endpoint and service, mirroring the structure of the GDC Variant Importer, for Asynchronous Publishing

**🐛 Fix Bugs in Async Publisher Service (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Add `await` before running validation and creating publish tasks to prevent starting empty publish queue
* Rearrange publish routes to put `/bulk` before `/:name` to prevent overlap

**📝 Update Documentation for Publish Router (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Include more details about the routes and their usage

**✨ Async Publishing (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Connect the UI to the new Asynchronous Publishing endpoints of the CMS, allowing the user to browse the CMS while publish operations are processing!
* Modify bulk publish route to accept lists of IDs as an alternative to model names
* Add new `PublishNotifications` system to UI, closely following the GDC Variant Import Notifications pattern

**💡 Update Comments** 

**⚡️ Optimize Async Bulk Publishes (https://github.com/nci-hcmi-catalog/portal/issues/1000)**
* Only update gene search indices after all models in a bulk publish are complete

### CODE REVIEW

File: ui/src/components/admin/ModelsManager/ModelColumns.js [hunk 1]
```diff
@@ -42,19 +43,50 @@ const selectedColumns = [
...
       return <SmallPill warning>{modelStatus.unpublishedChanges}</SmallPill>;
     case modelStatus.unpublished:
       return <SmallPill info>{modelStatus.unpublished}</SmallPill>;
     default:
-      return <SmallPill primary>{data.status}</SmallPill>;
+      return <SmallPill primary>{data}</SmallPill>;
   }
 };
 
+export const ModelStatusPillWithPublish = ({ name, data }) => {
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelColumns.js:59]: Enhanced version of the `ModelStatusPill` component which monitors publish progress.

Since we didn't want the added work of modifying the CMS and adding a new `publishing` state, but we still want the user to know if a model is currently being published, we can track if the model is being published and temporarily update the status here to `Publish in Progress`

File: ui/src/components/admin/ModelsManager/ModelColumns.js [hunk 2]
```diff
@@ -42,19 +43,50 @@ const selectedColumns = [
...
+  const { location, publishProgress } = useContext(NotificationsContext);
+  const { isPublishingModel, publishRunning } = usePublishNotifications();
+  const [status, setStatus] = useState(null);
+  const publishState = useRef();
+
+  // Force status pill to change to "Published" after single model in bulk publishes
+  // (prior to table data refresh that occurs after entire bulk publish finishes)
+  useEffect(() => {
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelColumns.js:67]: If a bulk publish completes while you are on the Model Management table, the data is refreshed to show the latest status of each model (Published, Changes Not Yet Published, Unpublished, etc)

However, with the introduction of the "Publish in Progress" status pull while publishing a model, we can get in the situation where the status pill goes from "Publish in Progress" back to "Changes Not Yet Published" if it finishes publishing before the rest of the bulk completes (and before the user navigates away/refreshes).

This useEffect hook forces the pill to go from "Publish in Progress" to "Published" if the publish completes successfully while the bulk is still ongoing. Note that this is only necessary on the Model Management table, not on the Edit Model page, hence the use of `location.pathname` here.

File: ui/src/components/admin/ModelsManager/ModelManagerController.js [hunk 3]
```diff
@@ -115,350 +120,356 @@ Model table state transitions for each action:
...
-            const [dataResponse, countResponse] = await loadData(baseUrl, state);
-            setState(() => ({
+    {({
+      appendNotification,
+      importProgress,
+      setImportProgress,
+      publishProgress,
+      setPublishProgress,
+    }) => (
+      <PublishNotificationsContext.Consumer>
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelManagerController.js:130]: Unfortunately the `ModelManagerController` isn't using hooks, and due to time constraints I didn't have the bandwidth to convert it while working on this. So, had to wrap the whole thing in  `<PublishNotificationsContext.Consumer>` here, and most of the changes to this file are just indentation/spacing changes as a result.

I've commented around the blocks which are actual changes, the rest can be ignored.

File: ui/src/components/admin/ModelsManager/ModelManagerController.js [hunk 4]
```diff
@@ -115,350 +120,356 @@ Model table state transitions for each action:
...
                       });
-                    }
-                  })
-                  .catch(async error => {
-                    await appendNotification({
-                      type: NOTIFICATION_TYPES.ERROR,
-                      message: 'Bulk Import of Research Somatic Variants Failed.',
-                      details: error.response ? error.response.data.error.message : error.message,
+                  },
+                  bulkPublish: async () => {
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelManagerController.js:324]: UI bulk publish function changes begin here (line 324) and go to line 354.

Removed locking the UI for synchronous publishing, replaced with call to async publisher.

File: ui/src/components/admin/ModelsManager/ModelManagerController.js [hunk 5]
```diff
@@ -115,350 +120,356 @@ Model table state transitions for each action:
...
-                  isLoading: true,
-                });
-
-                singleAction('publish', name)
-                  .then(() => loadData(baseUrl, state))
-                  .then(async ([dataResponse, countResponse]) => {
+                    await publishBulk(state.selection)
+                      .then(async response => {
+                        if (response.data.success) {
+                          await setPublishProgress({
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelManagerController.js:336]: If bulk publish is initiated successfully, we set `running` in the UI publishProgress to `true` which triggers the polling of publish progress to start (and displays the "Publishing:..." progress banner)

File: ui/src/components/admin/ModelsManager/ModelManagerController.js [hunk 6]
```diff
@@ -115,350 +120,356 @@ Model table state transitions for each action:
...
+                    appendNotification,
+                  }),
+                  bulkDelete: bulkActionCreator({
+                    action: 'delete',
+                    baseUrl,
+                    state,
+                    setState,
+                    appendNotification,
+                  }),
+                  publishOne: async name => {
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelManagerController.js:369]: `publishOne` (individual publish from Model Management table) has also been updated. It no longer locks the UI with `isLoading`, and instead calls the async publish function and creates an individual publish notification for that model.

File: ui/src/components/admin/ModelsManager/ModelManagerController.js [hunk 7]
```diff
@@ -115,350 +120,356 @@ Model table state transitions for each action:
...
+                    await publish(name)
+                      .then(async () => {
+                        await addPublishNotification(name);
+                      })
+                      .catch(async error => {
+                        const data = error.response ? error.response.data : error;
+                        showErrorPublishNotification(name, data);
+                      });
+                  },
+                  unpublishOne: async name => {
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelManagerController.js:379]: Any other changes to this file from here onward are just indentation/spacing changes as a result of wrapping this in another context consumer.

File: ui/src/components/admin/ModelsManager/ModelsManager.js [hunk 8]
```diff
@@ -20,7 +21,9 @@ import { AdminModalStyle } from 'theme/adminModalStyles';
 import config from '../config';
 import { BULK_UPLOAD_TYPES } from 'utils/constants';
 
-const content = () => {
+const ModelsManager = () => {
+  const { publishRunning } = usePublishNotifications();
```
Comment (mistryrn) [inline ui/src/components/admin/ModelsManager/ModelsManager.js:25]: Disable "Check for GDC Variants" and "Add Bulk" while a publish is running.

File: ui/src/components/admin/Notifications/NotificationToaster.js [hunk 9]
```diff
@@ -103,15 +105,20 @@ const renderIcon = type => {
 };
 
 const NotificationToaster = () => {
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/NotificationToaster.js:107]: Changes here are just:
- duplicate `import_____`
- replace `import` with `publish`

To extend the NotificationsToaster to support and display the progress banner for publish progress

File: ui/src/components/admin/Notifications/NotificationsController.js [hunk 10]
```diff
@@ -10,9 +10,13 @@ const NotificationsProvider = ({ location, children }) => {
   // only used for imports initiated from individual model pages
   const [importNotifications, setImportNotifications] = useState([]);
   // bulk variant import progress
-  const [importProgress, setImportProgress] = useState(DEFAULT_IMPORT_PROGRESS);
+  const [importProgress, setImportProgress] = useState(DEFAULT_PROGRESS_QUEUES);
   // models which failed variant import with nonactionable errors
   const [nonactionableImports, setNonactionableImports] = useState(DEFAULT_NONACTIONABLE_IMPORTS);
+  // only used for publishes initiated from individual model pages
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/NotificationsController.js:16]: Similar to the `NotificationsToaster` changes, we extend NotificationsController to support and display publish notifications.

File: ui/src/components/admin/Notifications/ProgressBanner.js [hunk 11]
```diff
@@ -216,7 +216,7 @@ const ProgressBanner = ({ renderIcon }) => {
   return (
     <Notification type={getProgressBannerType()}>
       {renderIcon(getProgressBannerType())}
-      <Col>
+      <Col style={{ maxWidth: '50%', width: '100%' }}>
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/ProgressBanner.js:219]: Message details are quite lengthy for Publish Progress, so we add a max width to make the text wrap sooner and give the rest of the components (progress bar, stop button) room to breathe.

File: ui/src/components/admin/Notifications/PublishNotifications.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+import NOTIFICATION_TYPES from './NotificationTypes';
+import {
+  acknowledgePublishStatus,
+  checkPublishStatus,
+} from 'components/admin/Model/actions/Publish';
+import { PUBLISH_ERRORS, PUBLISH_TYPES } from 'utils/constants';
+
+export const PublishNotificationsContext = React.createContext([{}, () => {}]);
+
+export const PublishNotificationsProvider = ({ children }) => {
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/PublishNotifications.js:12]: Modelled after [ui/src/components/admin/Notifications/GenomicVariantImportNotifications.js](https://github.com/nci-hcmi-catalog/portal/blob/develop/ui/src/components/admin/Notifications/GenomicVariantImportNotifications.js), but with all the GDC-related code removed, import error code removed, and converted to a Context so that `ModelsManagerController` can use it.

Comment (mistryrn) [reply ui/src/components/admin/Notifications/PublishNotifications.js:12]: Lots of duplicated/similar code between the linked file above and here where the only change is `import` -> `publish`.

A refactor could probably save a lot of duplication, but wasn't in the cards here due to time constraints. Possibility for a future ticket though.

File: ui/src/components/admin/Notifications/PublishProgress.js [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+} = base;
+
+const BulkPublishState = {
+  complete: 'COMPLETE',
+  stopped: 'STOPPED',
+  publishing: 'PUBLISHING',
+  off: 'OFF',
+};
+
+const PublishProgress = ({ renderIcon }) => {
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/PublishProgress.js:48]: Progress banner for publish progress. Very similar code to the import [ProgressBanner](https://github.com/nci-hcmi-catalog/portal/blob/develop/ui/src/components/admin/Notifications/ProgressBanner.js).

Pretty much the same thing as the aforementioned ProgressBanner, but publish-specific.

Definitely a future ticket/refactor where we make a generic progress banner component which takes in the status, notifications, labels, etc. as props. But again, time constraints 😓

**Review by mistryrn — COMMENTED** [review]
In lieu of a detailed PR description, I've gone through each file and added PR comments to help you in your review :)

File: ui/src/components/admin/Notifications/PublishNotifications.js [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+    const failedPublishes = failed || [];
+
+    failedPublishes.forEach(failedPublish => {
+      const modelName = failedPublish.modelName;
+
+      if (failedPublish.publishType === PUBLISH_TYPES.individual) {
+        removePublishNotification(modelName);
+      }
+
+      showErrorPublishNotification(modelName, failedPublish);
```
Comment (mistryrn) [inline ui/src/components/admin/Notifications/PublishNotifications.js:268]: A big difference between publish notifications and import notifications is that publish has no concept of "bulk" errors. For publish, the errors are all equal and we simply show them here, but for import there is a lot of extra logic surrounding whether the import was bulk or not.

File: cms/src/services/publish/Publisher.js [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+      }
+    }
+
+    if (running && queue.length > 0) {
+      run();
+    } else {
+      // Update gene search indices after bulk publish completes
+      await updateBulkGeneSearchIndicies();
+      pause();
+    }
```
Comment (joneubank) [inline cms/src/services/publish/Publisher.js:454]: 👍 
nice way to the bulk gene indexing is only run once at the end.

**Review by joneubank — APPROVED** [review]
